File: 2026/FF26-07abc.dart
import 'dart:io';
import 'dart:math';

void main() {
  final String input = File('input.txt').readAsStringSync();
  final (route, sushi) = process(input);

  print('Part 1 answer: ${part1(route, sushi)}');
  print('Part 2 answer: ${part2(route, sushi)}');
  print('Part 3 answer: ${part3(route, sushi)}');
}

int part1(List<Direction> route, List<Point> sushi) {
  Point snake = Point(0, 0);
  int i = 0;
  for (final Direction dir in route.sublist(0, route.length ~/ 2)) {
    snake.step(dir);
    if (snake == sushi[i]) i += 1;
  }

  return i;
}

int part2(List<Direction> route, List<Point> sushi) {
  List<Point> snake = [Point(0, 0)];
  int i = 0;
  for (final Direction dir in route) {
    snake.insert(0, snake.first.next(dir));
    if (snake.first == sushi[i]) i += 1;
    else snake.removeLast();

    if (snake.sublist(1).contains(snake.first)) return snake.length;
  }

  // Unreachable with valid inputs
  exit(1);
}

int part3(List<Direction> route, List<Point> sushi) {
  List<Point> snake = [Point(0, 0)];
  int i = 0;
  int ateSelf = 0;
  for (final Direction dir in route) {
    snake.insert(0, snake.first.next(dir));
    if (i < sushi.length && snake.first == sushi[i]) i += 1;
    else snake.removeLast();

    int index = snake.lastIndexOf(snake.first);
    if (index > 0) {
      snake.removeAfter(index - 1);
      ateSelf += 1;
    }
  }

  return snake.length * ateSelf;
}

(List<Direction>, List<Point>) process(String input) {
  final List<String> lines = input.lines().toList();

  final List<Direction> route = lines[0]
    .split('')
    .map(Direction.fromString)
    .toList();

  List<Point> sushi = [];
  for (final String line in lines.sublist(1)) {
    if (line.isEmpty) continue;

    final (a, b) = line.splitOnce(',')!;
    sushi.add(Point(int.parse(a), int.parse(b)));
  }

  return (route, sushi);
}

extension on List {
  void removeAfter(int start) {
    this.removeRange(start, this.length);
  }
}

extension on String {
  Iterable<String> lines() =>
    this.split('\n');

  (String, String)? splitOnce(String delimiter) {
    int i = this.indexOf(delimiter);
    if (i == null) return null;

    return (this.substring(0, i), this.substring(i+1));
  }
}

class Point {
  int x, y;

  Point(this.x, this.y);
  Point.clone(Point p) : this(p.x, p.y);

  Point next(Direction dir) =>
    this + switch (dir) {
      Direction.up    => Point( 0,  1),
      Direction.right => Point( 1,  0),
      Direction.down  => Point( 0, -1),
      Direction.left  => Point(-1,  0),
    };

  void step(Direction dir) {
    Point next = this.next(dir);

    this.x = next.x;
    this.y = next.y;
  }

  @override
  bool operator==(Object other) =>
    other is Point
    && this.x == other.x
    && this.y == other.y;

  @override
  int get hashCode =>
    Object.hash(x, y);

  Point operator+(Point other) =>
    Point(this.x + other.x, this.y + other.y);
}

enum Direction {
  up, right, down, left;

  static Direction fromString(String s) =>
    switch (s) {
      '^' => up,
      '>' => right,
      'v' => down,
      '<' => left,
      _ => exit(1),
    };
}

File: 2026/FF26-07abc.java
import java.util.Scanner;
import java.util.List;
import java.util.ArrayList;

public class Main {
  public static void main(String[] args) {
    Pair<List<Direction>, List<Point>> input = getInput();
    List<Direction> route = input.a;
    List<Point> sushi     = input.b;

    System.out.printf("Part 1 answer: %d\n", part1(route, sushi));
    System.out.printf("Part 2 answer: %d\n", part2(route, sushi));
    System.out.printf("Part 3 answer: %d\n", part3(route, sushi));
  }

  static int part1(List<Direction> route, List<Point> sushi) {
    Point snake = new Point(0, 0);

    int i = 0;
    for (Direction dir : route.subList(0, route.size()/2)) {
      snake.step(dir);
      if (snake.equals(sushi.get(i))) i += 1;
    }

    return i;
  }

  static int part2(List<Direction> route, List<Point> sushi) {
    List<Point> snake = new ArrayList<>();
    snake.add(new Point(0, 0));

    int i = 0;
    for (Direction dir : route) {
      snake.add(0, snake.get(0).next(dir));

      if (snake.get(0).equals(sushi.get(i))) i += 1;
      else snake.remove(snake.size() - 1);

      if (snake.subList(1, snake.size()).contains(snake.get(0))) return snake.size();
    }

    throw new RuntimeException();
  }

  static int part3(List<Direction> route, List<Point> sushi) {
    List<Point> snake = new ArrayList<>();
    snake.add(new Point(0, 0));

    int ateSelf = 0;
    int i = 0;
    for (Direction dir : route) {
      snake.add(0, snake.get(0).next(dir));

      if (i < sushi.size() && snake.get(0).equals(sushi.get(i))) i += 1;
      else snake.remove(snake.size() - 1);

      int index = snake.lastIndexOf(snake.get(0));
      if (index > 0) {
        snake = snake.subList(0, index-1);
        ateSelf += 1;
      }
    }

    return snake.size() * ateSelf;
  }

  static Pair<List<Direction>, List<Point>> getInput() {
    Scanner sc = new Scanner(System.in);

    String line = sc.nextLine();
    List<Direction> route = new ArrayList<>();
    for (char c : line.toCharArray()) {
      route.add(Direction.fromChar(c));
    }

    List<Point> sushi = new ArrayList<>();
    while (sc.hasNextLine()) {
      line = sc.nextLine();
      if (line.length() == 0) continue;

      String[] temp = line.split(",");
      sushi.add(new Point(Integer.parseInt(temp[0]), Integer.parseInt(temp[1])));
    }

    return new Pair<>(route, sushi);
  }
}

class Point {
  int x, y;

  public Point(int x, int y) {
    this.x = x;
    this.y = y;
  }

  public Point(Point p) {
    this.x = p.x;
    this.y = p.y;
  }

  public Point next(Direction dir) {
    return this.add(Point.fromDirection(dir));
  }

  public void step(Direction dir) {
    Point n = this.next(dir);
    this.x = n.x;
    this.y = n.y;
  }

  public static Point fromDirection(Direction dir) {
    switch (dir) {
    case UP:
      return new Point( 0,  1);
    case RIGHT:
      return new Point( 1,  0);
    case DOWN:
      return new Point( 0, -1);
    case LEFT:
      return new Point(-1,  0);
    default:
      throw new RuntimeException();
    }
  }

  public Point add(Point other) {
    return new Point(this.x + other.x, this.y + other.y);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof Point)) return false;

    Point other = (Point) o;
    return this.x == other.x && this.y == other.y;
  }
}

enum Direction {
  UP,
  RIGHT,
  DOWN,
  LEFT;

  public static Direction fromChar(char c) {
    switch (c) {
    case '^':
      return Direction.UP;
    case '>':
      return Direction.RIGHT;
    case 'v':
      return Direction.DOWN;
    case '<':
      return Direction.LEFT;
    default:
      throw new RuntimeException();
    }
  }
}

class Pair<A, B> {
  final A a;
  final B b;

  public Pair(A a, B b) {
    this.a = a;
    this.b = b;
  }
}